pkg/builder/builder.go

```diff
@@ -0,0 +1,540 @@
+	if err != nil {
+		return fmt.Errorf("failed to encode values: %v", err)
+	}
+	ctx.Values = &chart.Config{Raw: s}
+	return nil
+}
+
+func archiveSrc(ctx *Context) error {
+	contextDir, relDockerfile, err := build.GetContextFromLocalDir(ctx.AppDir, "")
+	if err != nil {
+		return fmt.Errorf("unable to prepare docker context: %s", err)
+	}
+	// canonicalize dockerfile name to a platform-independent one
+	relDockerfile, err = archive.CanonicalTarNameForPath(relDockerfile)
+	if err != nil {
+		return fmt.Errorf("cannot canonicalize dockerfile path %s: %v", relDockerfile, err)
+	}
+	f, err := os.Open(filepath.Join(contextDir, ".dockerignore"))
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	defer f.Close()
+
+	var excludes []string
+	if err == nil {
+		excludes, err = dockerignore.ReadAll(f)
+		if err != nil {
+			return err
+		}
+	}
+
+	// do not include the chart directory. That will be packaged separately.
+	excludes = append(excludes, filepath.Join(contextDir, "chart"))
+	if err := build.ValidateContextDirectory(contextDir, excludes); err != nil {
+		return fmt.Errorf("error checking docker context: '%s'", err)
+	}
+
+	// If .dockerignore mentions .dockerignore or the Dockerfile
+	// then make sure we send both files over to the daemon
+	// because Dockerfile is, obviously, needed no matter what, and
+	// .dockerignore is needed to know if either one needs to be
+	// removed. The daemon will remove them for us, if needed, after it
+	// parses the Dockerfile. Ignore errors here, as they will have been
+	// caught by validateContextDirectory above.
+	var includes = []string{"."}
+	keepThem1, _ := fileutils.Matches(".dockerignore", excludes)
+	keepThem2, _ := fileutils.Matches(relDockerfile, excludes)
+	if keepThem1 || keepThem2 {
+		includes = append(includes, ".dockerignore", relDockerfile)
+	}
+
+	logrus.Debugf("INCLUDES: %v", includes)
+	logrus.Debugf("EXCLUDES: %v", excludes)
+	rc, err := archive.TarWithOptions(contextDir, &archive.TarOptions{
+		Compression:     archive.Gzip,
+		ExcludePatterns: excludes,
+		IncludeFiles:    includes,
+	})
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+
+	var b bytes.Buffer
+	if _, err := io.Copy(&b, rc); err != nil {
+		return err
+	}
+	ctx.SrcName = "build.tar.gz"
+	ctx.Archive = b.Bytes()
+	return nil
+}
+
+// Up handles incoming draft up requests and returns a stream of summaries or error.
+func (b *Builder) Up(ctx context.Context, bctx *Context) <-chan *Summary {
+	ch := make(chan *Summary, 1)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		var (
+			buf = new(bytes.Buffer)
+			app *AppContext
+			err error
+		)
+		defer func() {
+			b.finish(app, buf)
+			wg.Done()
+		}()
+		if app, err = newAppContext(b, bctx, buf); err != nil {
+			fmt.Printf("buildApp: error creating app context: %v\n", err)
+			return
+		}
+		if err := b.buildImg(ctx, app, ch); err != nil {
+			fmt.Printf("buildApp: buildImg error: %v\n", err)
+			return
+		}
+		if app.ctx.Env.Registry != "" {
```

Comment:
If we are not on Minikube and the registry is not set, this will skip the pushing step and will try to release the chart without the image being pushed to a registry.

The Draft output is `SUCCESS`, but it actually silently fails as it cannot find the image.

Reply:
oops

Reply:
Sample output:

```
Draft Up Started: 'example-go'
example-go: Building Docker Image: SUCCESS ⚓  (5.0139s)
example-go: Releasing Application: SUCCESS ⚓  (0.9168s)
example-go: Build ID: 01C950XBFNB14Y7CKQ90BHRRN9
Inspect the logs with `draft logs 01C950XBFNB14Y7CKQ90BHRRN9`
```

Reply:
This is silently intentional. We have no way of detecting whether or not we are actually running against a non-minikube cluster. I've made a note of it in both the getting started guide as well as the reference docs, though I don't really see a way around this.

Any suggestions?

Reply:
The hack I implemented long ago in `draft init` mightt work here (relying on the kubernetes config's current-context name to be "minikube", but it was more or less a crutch and probably won't hold up forever.

Reply:
Would it be fair to inform the user that no registry was set before starting the build?

Reply:
The least we can do I think is outputting that the push step is skipped because there is no registry set in the config.

It would at least inform the user why.

Another thing that I have in mind - can we detect if it's Minikube by the master URL when using the Kubernetes client? It's surely localhost when Minikube, right?

Reply:
I don't like the "minikube" magic string solution, as I now have five different minikube profiles for different configurations, as you can imagine. I prefer to a) inform the person that no registry was set, so if it's not running against minikube it will skip the push activity and report success and yet fail; and b) doc the heck out of it.

However, another option might be to have the default be to push, but have the --skip-push flag such that when you know you're local, do this to skip push? Dunno. Whatchya think.

Reply:
Oh, GitHub refresh...

Reply:
@squillace I considered the --skip-push flag, but that means minikube users now need to run `draft up --skip-push` all the time. It's a little more cumbersome for an "optional" feature flag, so I wanted to make the default experience sane.

I've made a change in 75fd33f so it'll display a warning, as well as a way to disable that warning if you know what you're doing. Sounds good?

Reply:
Let's try it for a bit, and if it's an issue we can identify a better ux. at this point, it's really a "did you remember to specify a registry if you're working against the cloud?" problem. It's not really a draft problem.